src/Compiler/Checking/import.fs

```diff
@@ -43,48 +45,101 @@ type AssemblyLoader =
     /// Record a root for a [<Generate>] type to help guide static linking & type relocation
     abstract RecordGeneratedTypeRoot : ProviderGeneratedType -> unit
 #endif
-        
+
+[<Struct; NoComparison>]
+type CanCoerce =
+    | CanCoerce
+    | NoCoerce
+
+type [<Struct; NoComparison; CustomEquality>] TTypeCacheKey =
+
+    val ty1: TType
+    val ty2: TType
+    val canCoerce: CanCoerce
+    val tcGlobals: TcGlobals
+
+    new (ty1, ty2, canCoerce, tcGlobals) =
+        { ty1 = ty1; ty2 = ty2; canCoerce = canCoerce; tcGlobals = tcGlobals }
+
+    interface System.IEquatable<TTypeCacheKey> with
+        member this.Equals other =
+            let ty1 = this.ty1
+            let ty2 = this.ty2
+            let otherTy1 = stripTyEqns this.tcGlobals other.ty1
```

Comment:
I'm not 100% sure we need to be adding separate type here, unless we do it systematically everywhere across the whole codebase, otherwise its existence will be confusing for people. We can strip when we do type cosntruction, and don't strip it only on equality.

Reply:
The callsite (typeFeasibleSubsumes..) already does stripping, so even type constructor would do double work.
But I can do with a dual constructor - one that will strip, and an explicit named static method `TTypeCacheKey.FromStrippedTypes(sty1,sty2,..)` that will not do a repetitive strip.

Reply:
I suppose, this is minuscule though, I doubt this will affect anything, especially in this scenario.